### DISCUSSION TITLE
Dev Approach

### DISCUSSION THREAD

**Discussion by cdnninja** [audiconnect/audi_connect_ha#273] [General]
Hi All, 

Wanted to started a thread talking about the future of this integration.      I have merge access to this repo but not full admin.  

I am hoping to coordinate our efforts as lots of good work is going into this.  Wanting to hear thoughts on if others working agree or would like to fork into new repos.   Here is my hope:

1.  Convert this repo to a org repo.  So audi_connect_ha/audi_connect_ha.  If converted it auto redirects hacs.  @arjenvrh  would need to give admin or do this for us.   I would propose a few of us get owner admin to the new org.  
2. Use PRs to work together to improve this and resolve the issues. 
3. Include release version control much like what I have done in kia_uvo repo that I am moving on from. 

If the preference is to fork and move on from here please let me know since I do understand. 

Lets keep this issue focused around dev approach and contributions and not issues. 

@t0bias-r @gac3k @DivingUp

**Comment by t0bias-r** [comment]
What is the goal?
To be as fast as possible or to have longterm support with stable code?

There was a web API change. I would say that it's ok when it may take a few days until everything works again. And e.g. an AdBlue sensor value is not that super urgent in my opinion.
I would place more value on good and stable solutions, PRs that are really discussed, code that is reviewed and approved instead of clicking on the merge button as quick as possible.

**Comment by DivingUp** [comment]
the AdBlue patch had to be merged, because without this patch the application did not work for all vehicles without AdBlue... you just have to follow the comments...
but it also seems to be a bit of mimimi here...

**Comment by Kolbi** [comment]
I like more the approach from @cyr-ius : https://github.com/cyr-ius/hass-audiconnect // https://github.com/cyr-ius/audiconnectpy as it splits HASS & audiconnectpy functionalities - which seems to be coming standard in HASSIO development. Sadly he seems sometimes not to be very active. :)
He also integrated different API levels for the different vehicles & functionalities. But this topic may be otudated with the new APIs.

**Comment by cdnninja** [comment]
> What is the goal?
> 
> To be as fast as possible or to have longterm support with stable code?
> 
> 
> 
> There was a web API change. I would say that it's ok when it may take a few days until everything works again. And e.g. an AdBlue sensor value is not that super urgent in my opinion.
> 
> I would place more value on good and stable solutions, PRs that are really discussed, code that is reviewed and approved instead of clicking on the merge button as quick as possible.



Thank you for the reply. I'm open to that and agree.  Purely merged to get things working. 

My impression code in general could use cleanup. I'll slow merging and I request both of you provide reviews to PRs. 

I only have merge access. Not admin unfortunately.

**Comment by cdnninja** [comment]
> I like more the approach from @cyr-ius : [cyr-ius/hass-audiconnect](https://github.com/cyr-ius/hass-audiconnect) // [cyr-ius/audiconnectpy](https://github.com/cyr-ius/audiconnectpy) as it splits HASS & audiconnectpy functionalities - which seems to be coming standard in HASSIO development. Sadly he seems sometimes not to be very active. :) He also integrated different API levels for the different vehicles & functionalities. But this topic may be otudated with the new APIs.

This is how we built out version 2 of the kia integration in my past life.  It supported others in automation as well because of that.

**Comment by cdnninja** [comment]
@arjenvrh  wanted to ping you again to see if we could get admin to convert this to a shared repo.  With your permission and support.  This repo is starting to take life so hoping to keep that going.

**Comment by arjenvrh** [comment]
@cdnninja I would be more than happy to convert this repo to make collaboration/maintaining easier. I'm not sure how to give you admin access to this repo. Or do you mean that I create an organization, shift this repo to this organization and add you to the organisation?

**Comment by t0bias-r** [comment]
@cdnninja @DivingUp @arjenvrh
When the repo is moved to an organization, the URL will change.
And when it is ok for us and the endusers, that the URL changes, then you @cdnninja could make a new fork within your account and we maintain that fork together. Then @arjenvrh does not have to give admin rights.

**Comment by arjenvrh** [comment]
Just let me know what you all prefer. I can add collaborators to this repo, create an org, or do nothing and we all move to a fork/different URL. When changing the URL would be ok, the org-way would perhaps have the advantage that when the new maintainer later doesn't have the time anymore to maintain the project, others could take over without changing the url again.

**Comment by cdnninja** [comment]
> @cdnninja @DivingUp @arjenvrh
> 
> When the repo is moved to an organization, the URL will change.
> 
> And when it is ok for us and the endusers, that the URL changes, then you @cdnninja could make a new fork within your account and we maintain that fork together. Then @arjenvrh does not have to give admin rights.

GitHub does built in redirects if we use this one and convert to org. So HACS will move users over.  We did this for Kia_uvo and worked great.

**Comment by cdnninja** [comment]
> Just let me know what you all prefer. I can add collaborators to this repo, create an org, or do nothing and we all move to a fork/different URL. When changing the URL would be ok, the org-way would perhaps have the advantage that when the new maintainer later doesn't have the time anymore to maintain the project, others could take over without changing the url again.

Just checked. I think where you set me as a collaborator just change that to "owner" and it will allow me to make the changes.    Thank you for all your work on this project you paved the way. @arjenvrh

**Comment by Kolbi** [comment]
Bespite that discussion do we have a full list of actual API calls from the app?

And maybe even with sample data?

**Comment by t0bias-r** [comment]
> Bespite that discussion do we have a full list of actual API calls from the app?
> 
> And maybe even with sample data?

I can capture the network traffic of the Android myAudi app and store the decrypted traffic in a HAR file. This would be for a Q8 e-tron.
Because I don't have much time, I can't code it myself. But I could provide such HAR files privately.
I doubt there is a public specification.

**Comment by TheDelta** [comment]
Hi, been following this repo for a while and got active because of the new API changes.
I forked `https://github.com/cyr-ius/audiconnectpy` (https://github.com/TheDelta/audiconnectpy) and fixed current API issues to get at least vehicle information back.

Would love to contribute as well and get a fully working home assistant integration back for a broader set of Audi vehicles (I could contribute data from an Audi RS3 2023 / MIB3 platform)

My initial question would be just what starting base should be used? This repo or the ones of @cyr-ius ?

**Comment by Ingrimmsch93** [comment]
So as i understand this repo will probably be moved to an organisation, so it is easier to maintain. With @cyr-ius permission, we could include his work to combine both and start with arjenvrh version. For me it looks like a good start for a "new" audi integration.
I am pretty new, but very interested to learn and contribute. Time is also a topic for me. Currently i can spend some time.

I can contribute data from an Audi A3 2023 TFSIe. @t0bias-r could you send me an HAR file from you or from my account? I think this would help me/us a lot to see whats actually available. I also would check, if i can anonymize the HAR file and share it here in any way.

**Comment by arjenvrh** [comment]
@cdnninja I cannot find an option to make you owner. I created an organization audiconnect for you and invited you as an owner.  Then I will move the repo to this organization. Is that ok?

**Comment by Kolbi** [comment]
> Hi, been following this repo for a while and got active because of the new API changes. I forked `https://github.com/cyr-ius/audiconnectpy` (https://github.com/TheDelta/audiconnectpy) and fixed current API issues to get at least vehicle information back.
> 
> Would love to contribute as well and get a fully working home assistant integration back for a broader set of Audi vehicles (I could contribute data from an Audi RS3 2023 / MIB3 platform)
> 
> My initial question would be just what starting base should be used? This repo or the ones of @cyr-ius ?

The architecure from @cyr-ius looks better for me splitting hassio frontend and api backend. But he seems not to be very active and here we have some active people already :) You are not living so far away from me.

**Comment by tieskuh** [comment]
> Hi, been following this repo for a while and got active because of the new API changes. I forked `https://github.com/cyr-ius/audiconnectpy` (https://github.com/TheDelta/audiconnectpy) and fixed current API issues to get at least vehicle information back.
> 
> Would love to contribute as well and get a fully working home assistant integration back for a broader set of Audi vehicles (I could contribute data from an Audi RS3 2023 / MIB3 platform)
> 
> My initial question would be just what starting base should be used? This repo or the ones of @cyr-ius ?

I noticed you didn't create a pull request in cyr-ius' repo with your fixes. Maybe you could do that to prevent him from doing the same changes over.

**Comment by tieskuh** [comment]
> > Hi, been following this repo for a while and got active because of the new API changes. I forked `https://github.com/cyr-ius/audiconnectpy` (https://github.com/TheDelta/audiconnectpy) and fixed current API issues to get at least vehicle information back.
> > Would love to contribute as well and get a fully working home assistant integration back for a broader set of Audi vehicles (I could contribute data from an Audi RS3 2023 / MIB3 platform)
> > My initial question would be just what starting base should be used? This repo or the ones of @cyr-ius ?
> 
> The architecure from @cyr-ius looks better for me splitting hassio frontend and api backend. But he seems not to be very active and here we have some active people already :) You are not living so far away from me.

I have been using cyr-ius' repo for quite some time now. It has been very stable until now. The last few weeks he has not been very active indeed but before he responded very fast. Maybe he is busy at work or on holiday. I think it could be a good idea to reach out to him to discuss. He refactored the entire code. In arjenvrh's repo there are more active contributers, but maybe cyr-ius is open to collaboration as well.

**Comment by cdnninja** [comment]
> @cdnninja I cannot find an option to make you owner. I created an organization audiconnect for you and invited you as an owner. Then I will move the repo to this organization. Is that ok?

Yes that works.  Ready to go when you are!  

@tieskuh @cyr-ius yes I personally hope to be able to bring many together on this.   This repo seems to have the most users so thinking it should live on but could be refactored as wel. 

Stats from home assistant show    

 "audiconnect": {
        "total": 605,
        "versions": {
            "2021.1.1": 582,
            "1.5.1": 15,
            "2023.12.3": 8

2021.1.1 being this repo.  1.5.1 being cyr-ius I believe.

**Comment by arjenvrh** [comment]
I transferred the repo to the audiconnect organization. Redirect seems to be working. Thanks to all contributors for keeping the Audi integration alive!

**Reply by cdnninja** [reply]
Thank you for all the work you have done to take it this far!  Greatly appreciated!

**Comment by TheDelta** [comment]
> > Hi, been following this repo for a while and got active because of the new API changes. I forked `https://github.com/cyr-ius/audiconnectpy` (https://github.com/TheDelta/audiconnectpy) and fixed current API issues to get at least vehicle information back.
> > Would love to contribute as well and get a fully working home assistant integration back for a broader set of Audi vehicles (I could contribute data from an Audi RS3 2023 / MIB3 platform)
> > My initial question would be just what starting base should be used? This repo or the ones of @cyr-ius ?
> 
> I noticed you didn't create a pull request in cyr-ius' repo with your fixes. Maybe you could do that to prevent him from doing the same changes over.

Key issue is that I don't want to "waste" too much time on something which might be not the development direction and also it's not fully working with legacy yet. Could need some more love and more brain power. But I look into it how to continue. Top priority should be to get most if not all entities back asap now.

**Comment by cdnninja** [comment]
It is a very good point.  I don't know the best code base to start with.  This repo being tied to users HACS allows easy upgrades.  It can be switched to any code easily with permission of course. 

I don't know enough on each repo to comment on the better code base though.  Nor am I a professional dev.  Just a hobby.  

Another option is patch this up a bit to keep it working.  Then work on v2 as a branch here.  Once ready merge to master.  Can use the best approach from multiple sources.   However that isn't a small task. 

I do like splitting out a separate python library for the api vs HA.  That is what I used in the paste project and it helped solve many issues.

**Comment by cdnninja** [comment]
I have enabled discussions and converted this to one.   This should allow us to have a few topics around dev work.  I enabled releases and created a discussion to talk about it if not working for everyone.

**Comment by TheDelta** [comment]
Thanks to all involved, great to see a v1 and working integration again 🥳
Went pretty fast compared to other projects ❤️

**Comment by t0bias-r** [comment]
> Another option is patch this up a bit to keep it working. Then work on v2 as a branch here. Once ready merge to master. Can use the best approach from multiple sources. However that isn't a small task.
> 
> I do like splitting out a separate python library for the api vs HA. That is what I used in the paste project and it helped solve many issues.

Splitting Audi API and HA integration would be great. 👍
- Keeping existing `master` as legacy
- develop the API/HA split thing in a dedicated branch
- and when everything works, merge this dev-branch into master

I can contribute primarily to the API, but I am limited in time.

**Reply by cdnninja** [reply]
I like it.  Could you review https://github.com/Hyundai-Kia-Connect/kia_uvo/ and https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api to see if architecture makes sense to follow? It actually is very similar to @cyr-ius repo.  It followed this same path we talked about.  Coming from a v1 and converting to what you see today when it was ready for cutover. 

Would love to hear from @cyr-ius if he is interested in joining us.

**Reply by cdnninja** [reply]
@t0bias-r  If I set up a basic structure of the API and get it uploading to pypi would you have cycle to work on that?

**Reply by t0bias-r** [reply]
Sounds great, but in the next few weeks I have almost no time to work on this.

**Reply by Kolbi** [reply]
@cyr-ius seems to be active again in his repos.

**Comment by DivingUp** [comment]
In principle, I also think it's a good idea to make the split. But I'm also limited in terms of time... My focus at the moment is on the stability of the functions / customization of the app so that it continues to work / works again because I am also limited in time.

how do we want to do this with the owner structure of audiconnect?
do you also want to add us as owners or do you want to leave it as it is?
should we install a way for internal / non-public communication for consultation and discussion ?

Translated with DeepL.com (free version)

**Reply by cdnninja** [reply]
I think I granted you both org owners so you should have same access to me.   I believe having multiple people with access is key since we all have life and other priorities.  

Also open to discord or something if you prefer.

**Reply by DivingUp** [reply]
this is for me fine.

**Comment by Kolbi** [comment]
I activated pre-commit for all repositories for audiconnect: https://results.pre-commit.ci/repo/github/186347733

Because I guess this will help us to improve code. We can discuss (if needed) the https://github.com/audiconnect/audi_connect_ha/blob/master/.pre-commit-config.yaml

**Reply by cdnninja** [reply]
Thanks for doing this! Very cool tool!

**Comment by jeffkibuule** [comment]
Given all of the forks that exist, can we determine what the canonical repo/URL should be? It's difficult for the average person to determine which one to add to HACS because of the numerous suggested repos to add. Thanks!

**Reply by Kolbi** [reply]
Currently you don't need to add a custom repository to HACS as this repo is added to HACS per default: https://github.com/hacs/default/blob/ad1a70511ca978ec29216bf0cb4283e55a6b6387/integration#L75C3-L76C27 (yeah the old URL but should redirect to the new one).

**Reply by jeffkibuule** [reply]
Thanks for the reply
